File: Hard/123. Best Time to Buy and Sell Stock III.js
/*  This question is solved but by mathmatically */

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    var fb = Number.MIN_VALUE, sb = Number.MIN_VALUE;
    var fs = 0, ss = 0;
    for (let i = 0; i < prices.length; i++) {
        fb = Math.max(fb, -prices[i]);
        fs = Math.max(fs, parseInt(fb) + parseInt(prices[i]));

        sb = Math.max(sb, parseInt(fs) - parseInt(prices[i]));
        ss = Math.max(ss, parseInt(sb) + parseInt(prices[i]));
    }
    return ss;
};

console.log(maxProfit([3, 3, 5, 0, 0, 3, 1, 4]));

File: Hard/123. Best Time to Buy and Sell Stock III.java
class Solution {

    public int maxProfit(int[] prices) {
        int fb = Integer.MIN_VALUE, sb = Integer.MIN_VALUE;
        int fs = 0, ss = 0;

        for (int i = 0; i < prices.length; i++) {
            fb = Math.max(fb, -prices[i]);
            fs = Math.max(fs, fb + prices[i]);
            
            sb = Math.max(sb, fs - prices[i]);
            ss = Math.max(ss, sb + prices[i]);
        }
        return ss;
    }
}
